Add per-channel custom quantization levels (non-linear palettes)

ulalume opened the pull request:
## Summary

-   Adds optional per-channel custom level lists to `QualetizePlan`,
    allowing palettes to snap to arbitrary, non-linear steps (e.g., Sega
    Genesis 3-bit RGB levels), while keeping existing linear `-rgba:`
    depth as the default.
-   Extends the CLI with `-customlevels:` (slash-separated lists) so
    users can supply these levels without affecting existing workflows.
    Quantization remains order-independent and fully backward
    compatible.
-   No changes to the `Qualetize` API signature; existing callers
    continue to work as is.

## Motivation

-   Some systems (e.g., Sega Genesis / Mega Drive) use non-linear DAC
    steps for their "3-bit" RGB values (see forum reference:
    <https://gendev.spritesmind.net/forum/viewtopic.php?f=22&t=2188>).
    With 3-bit depth, the current linear quantizer outputs 0, 36, 73,
    ..., but the real hardware levels are:

    **0, 52, 87, 116, 144, 172, 206, 255**

-   Custom level support allows users to specify exact hardware steps or
    other specialized palettes without altering default behavior or
    `-rgba:` usage.

## Changes

-   **`QualetizePlan_t`**:
    -   Adds `CustomLevels[4]` and `CustomLevelCount[4]` (RGBA).
    -   If null/0, the quantizer falls back to the existing
        `ColourDepth` logic.
-   **Quantization**:
    -   Palette generation uses plan-aware quantization (`nearest`,
        `floor`, `ceil`) with order-independent behavior.
-   **CLI**:
    -   Adds `-customlevels:r.../g.../b.../a...`, where each channel
        list contains values from 0--255.
    -   Existing options remain unchanged.

## Testing

-   **Baseline compatibility**

        qualetize in.bmp out.bmp -npal:16 -cols:16 -rgba:3331

    Output should match current behavior.

-   **Custom levels example**

        qualetize in.bmp out.bmp -npal:1 -cols:16 -customlevels:0,52,87,116,144,172,206,255/0,52,87,116,144,172,206,255/0,52,87,116,144,172,206,255/0,255

    RGB values should snap to the Genesis hardware steps.

## Notes

-   The API surface remains unchanged; only the plan struct is extended.
-   Custom levels are optional and per-channel; alpha may continue using
    `ColourDepth` or its own list.
-   If this addition doesn't align with the project's goals, feel free
    to close --- I'm happy to keep it on my fork. Others may find the
    non-linear palette option useful 😃

ulalume committed:
Add support for custom quantization levels

Introduces per-channel custom quantization levels for RGBA in the QualetizePlan struct and implements parsing and application of these levels in the CLI and core quantization logic. This allows users to specify exact quantization values for each channel via the new -customlevels option, improving flexibility for specialized output requirements.

Aikku93 commented:
Thanks! I don't really have a use for this, but I'm happy to merge it in anyway. I'm personally still trying to figure out a better method of clustering the tiles... I even tried using ChatGPT 😂

I would suggest a few things, though (just let me know if you disagree with anything, I'm not too fussed):
* Assuming that ColourLevels[] is in ascending order, you can terminate early once you find the cutoff point (eg. for FLOOR mode: terminate once the next level is higher or end-of-list; for CEIL mode, terminate when the next level is higher or end-of-list and add 1 if not end of list; for ROUND mode, terminate when the next level is higher or end-of-list, and increment by 1 if the next level's error is lower than the current one). Because all of these cases have a common "starting point", you can just fold it into a single search loop, and then just fix up CEIL and ROUND as needed at the end.
  * This does mean that the list needs to provided in ascending order by the user, or sorted, but that's trivial.
  * I guess you could use a binary search too, but that's probably overkill.
* Spacing is broken on lines 117-118 near `floorf()` and `ceilf()` in Qualetize.c
* Allowing a single ColourLevels[] list to be used for all channels could be beneficial and less annoying to use if the target system has the same curve for all channels. Suggested implementation would be to check for a completely empty list after the `/` separator (eg. `0,63,128,191,255///0,1` would duplicate the R channel list for G and B channels).

ulalume committed:
Fix indentation in QuantizeChannel_WithPlan function

ulalume committed:
Update help text for -customlevels option

Moved and expanded the description for the -customlevels option in the CLI help text. Added clarification that empty sections reuse the previous channel value.

ulalume committed:
Support empty channel reuse in -customlevels (copy previous list)

- Allow empty sections in -customlevels (e.g. r///a) to reuse the previous channel’s level list, reducing redundant input when all RGB share the same curve.
- Parser now accepts empty lists, copies the prior channel when empty, and keeps ownership tracking so shared lists aren’t double-freed.

ulalume committed:
Optimize custom level quantization with single-pass ascending search

- Assume custom level lists are ascending and use a single-pass scan to find floor/ceil/nearest, breaking as soon as the first value > v is seen.
- Floor/ceil are now early-exit on the cutoff; nearest reuses the same candidates without a full pass.
- Updated CLI help to state that custom level lists should be provided in ascending order.

ulalume commented:
Thanks for the review and suggestions!  
I’ve updated the PR with three improvements:

- Optimized custom level quantization using a single-pass ascending search
- Added support for empty channel reuse in -customlevels (copies previous list)
- Fixed indentation issues in QuantizeChannel_WithPlan

I hope these changes match your suggestions correctly… 🥹
